File: src/analise_semantica/intrinsics.py
"""Especificacao das funcoes intrinsecas suportadas."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class IntrinsicSpec:
    name: str
    arity: int | None
    return_type: str | None = None


INTRINSICS: dict[str, IntrinsicSpec] = {
    "MOD": IntrinsicSpec("MOD", 2, "INTEGER"),
    "INT": IntrinsicSpec("INT", 1, "INTEGER"),
    "REAL": IntrinsicSpec("REAL", 1, "REAL"),
    "FLOAT": IntrinsicSpec("FLOAT", 1, "REAL"),
    "ABS": IntrinsicSpec("ABS", 1, None),
    "SQRT": IntrinsicSpec("SQRT", 1, "REAL"),
    "MAX": IntrinsicSpec("MAX", 2, None),
    "MIN": IntrinsicSpec("MIN", 2, None),
}

File: tests/fixtures/intrinsics.f
PROGRAM INTRINS
       INTEGER I, J, K
       REAL R, S, T
       I = 3
       J = 7
       R = ABS(-4.0)
       S = SQRT(9.0)
       K = MAX(I, 2) + MIN(J, 5)
       T = FLOAT(K) / 2.0
       PRINT *, 'Resultados:', R, S, K, T
       END
